src/main/java/ch/njol/skript/effects/EffConnect.java

```diff
@@ -32,62 +13,94 @@
 	private Expression<Player> players;
-
-	@SuppressWarnings("null")
 	private Expression<String> server;
+	private Expression<Number> port;
+	private boolean transfer;
 
 	@Override
-	protected void execute(Event e) {
-		String server = this.server.getSingle(e);
-		Player[] players = this.players.stream(e)
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+		players = (Expression<Player>) exprs[0];
+		server = (Expression<String>) exprs[1];
+		port = (Expression<Number>) exprs[2];
+		transfer = matchedPattern == 1;
+		if (!TRANSFER_PACKET_EXISTS && transfer) {
+			Skript.error("Transferring players via IP is not available on this version.");
+			return false;
+		}
+		return true;
+	}
+
+	@Override
+	protected void execute(Event event) {
+		String server = this.server.getSingle(event);
+		Player[] players = this.players.stream(event)
 			.filter(Player::isOnline)
 			.toArray(Player[]::new);
-		if (server == null || players.length == 0)
+
+		if (players.length == 0)
 			return;
 
-		// the message channel is case sensitive so let's fix that
-		Utils.sendPluginMessage(players[0], BUNGEE_CHANNEL, r -> GET_SERVERS_CHANNEL.equals(r.readUTF()), GET_SERVERS_CHANNEL)
-			.thenAccept(response -> {
-				// for loop isn't as pretty a stream, but will be faster with tons of servers
-				for (String validServer : response.readUTF().split(", ")) {
-					if (validServer.equalsIgnoreCase(server)) {
-						for (Player player : players)
-							Utils.sendPluginMessage(player, BUNGEE_CHANNEL, CONNECT_CHANNEL, validServer);
-						break;
-					}
+		if (transfer) {
+			int port = 25565; // default port
+			if (this.port != null) {
+				Number portNum = this.port.getSingle(event);
+				if (portNum != null) {
+					port = portNum.intValue();
 				}
-			});
+			}
+			for (Player player : players) {
+				player.transfer(server, port);
+			}
+		} else {
+			// the message channel is case-sensitive, so let's fix that
+			Utils.sendPluginMessage(players[0], BUNGEE_CHANNEL, r -> GET_SERVERS_CHANNEL.equals(r.readUTF()), GET_SERVERS_CHANNEL)
+				.thenAccept(response -> {
+					// for loop isn't as pretty as a stream, but will be faster with tons of servers
+					for (String validServer : response.readUTF().split(", ")) {
+						if (validServer.equalsIgnoreCase(server)) {
+							for (Player player : players)
+								Utils.sendPluginMessage(player, BUNGEE_CHANNEL, CONNECT_CHANNEL, validServer);
+							break;
+						}
+					}
+				});
+		}
 	}
 
 	@Override
-	public String toString(@Nullable Event e, boolean debug) {
-		return "connect " + players.toString(e, debug) + " to " + server.toString(e, debug);
+	public String toString(@Nullable Event event, boolean debug) {
+		String portString = port != null ? (" and on port " + port.toString(event, debug)) : "";
+		return "connect or transfer " + players.toString(event, debug) + " to " + server.toString(event, debug) + portString;
```

Comment:
this should actually use the pattern the player used and not just half-ass it by saying `a or b`